### ISSUE TITLE
Serious usability issues prompting for credentials with an separate LFS server

### ISSUE THREAD

**Issue by rjbell4** [github/git-lfs#1121]
I have observed serious usability issues prompting for credentials, when using a separate LFS server.  In my case, I am using GitHub Enterprise, and a separate Artifactory server for LFS, though I don't believe those specifics would be necessary.  The usability issues are severe enough for me to consider the case "unusable" and a bug.

Once credentials are cached, everything is smooth.  However, if credentials have not been cached, prompting for them is basically usable.  The workaround for the situation is to ensure credentials are cached (or stored) via an "out-of-band" mechanism.  This isn't a good sell to potential Git LFS users, though. :disappointed: 

First, I appear to be prompted for the password once for each of the "ConcurrentTransfers" threads.  These prompts are all strung together on a single line, and I have to enter credentials over and over again.

Additionally, after entering a username, the prompt for a password appears, but is immediately overwritten with an update on the status.  Therefore, it's not clear that's there's even a prompt, much less what is required.

I've attached an image to attempt to demonstrate the usability issue in this situation.  My configuration for the example is simply:
```
credential.helper=cache
http.sslverify=false
lfs.concurrenttransfers=8
```

It is the output of ```git lfs fetch``` from a repository cloned with ```GIT_LFS_SKIP_SMUDGE=1```.  The progress being made is as I attempt to enter my username and password multiple times, but I'm not even clear what I'm being prompted for, so I eventually give up.  I don't want other users to go through the same thing!

![image](https://cloud.githubusercontent.com/assets/375968/14251722/503d9512-fa52-11e5-9b85-ca2b51584d52.png)

**Comment by technoweenie** [comment]
Thanks for the report. Looks like an issue with the progress tracker library. While it's definitely a bug worth fixing, your best bet will always be using a credential cacher.

**Comment by rjbell4** [comment]
@technoweenie I am using a credential cacher.  However, the problem comes when nothing is cached (yet) for the separate LFS server (Artifactory, in my cache).  It's less of an issue when your Git server and LFS server are one and the same, as your credentials probably are cached from Git (non-LFS) operations.  But in my case, they are separate, so credentials are cached separately.  The above operation is how the credentials would normally get cached in the first place.

As noted, once cached, this isn't a problem at all.  However, I have to resort to some "out of band" mechanism to get them cached in the first place, since the normal way of being prompted during a fetch isn't worked.  That's rather painful for many developers.

**Comment by sinbad** [comment]
I was looking at this on chat, and I can't see where we block workers before the 1st auth is confirmed in the batch route; it's still there in the old non-batch route but I can't find anything that would force only *one* worker to prompt for credentials if they're not cached (which they won't be in the case of a separately hosted git repo and lfs server). Unless I'm missing it this could be the cause - I generally use co-hosted git and lfs stores so I never see it.

**Comment by sinbad** [comment]
Ah, no I think I get it; the batch routine only has a single worker calling `Batch()` (inside `batchApiRoutine`) so any auth that occurs on the LFS API would be serialized. However if the host that is providing the LFS *storage* is different to the one implementing the LFS *API*, and the LFS *storage* server requires separate auth (not embedded in a token that's returned from the API as part of the URL), then the transfer workers would all prompt for auth in parallel.

This must be something quite specific to the way Artifactory is doing things, I haven't come across anyone else doing it like this; when the LFS store is separate usually auth is embedded in the transfer URL with something like JWT (as with S3 storage etc), rather than requiring another level of auth.

**Comment by technoweenie** [comment]
I don't think there's anything wrong with the way Artifactory is doing things here. 

I thought the transfer queue object had something that would block all of the concurrent goroutines until there was a successful auth. Maybe I'm wrong, or some http response from Artifactory is confusing the client?

**Comment by sinbad** [comment]
If the LFS API was implemented in Artifactory it wouldn't be a problem, because the API calls are serialized & would only prompt once. The issue as I interpret it is that the LFS API is on GHE but the store is on Artifactory? That's kind of like S3, but in that case you'd use a token generated by the LFS API to auth on S3 so you'd never have this issue.

The actual transfer workers that run in parallel on the results of the batch API are not blocked until successful auth.

**Comment by rjbell4** [comment]
@sinbad My lfs.ufs in .lfsconfig points directly at Artifactory, not GHE.  I believe that 100% of the LFS API is handled by Artifactory itself.

**Comment by sinbad** [comment]
In that case the only way I can see that happening is if the non-API part of the URL (the download) that the LFS client hits after making the batch API call requires separate auth (no token), and which doesn't match any credentials cached from the earlier Artifactory LFS Batch API call (which is serialized).

**Comment by rjbell4** [comment]
Attaching a run with GIT_TRACE=1 set.  I Ctrl-C'd immediately when prompted (under the presumption  that the prompt is already encountered at that point).  Let me know if some alternative form of debugging would be more helpful.
[fetch-trace.txt](https://github.com/github/git-lfs/files/202795/fetch-trace.txt)

**Comment by sinbad** [comment]
Well, that's my initial theory out of the window then :laughing: The prefixes on the endpoints are the same for API & download:

```
http://afeossand1.cec.lab.emc.com/artifactory/api/lfs/upc_git_lfs/objects/batch
http://afeossand1.cec.lab.emc.com/artifactory/upc_git_lfs/objects/0d/8c/0d8c6d7bc3c7a3768231cf16b23a09eb5d9635fd1a0303932b47ad548d22fe1c
```
The API call clearly worked fine, and the URL prefix is the same so the creds should have been available, so maybe it's something simpler like how we're retrieving/supplying credentials for the download calls.  There's no embedded auth in the download URLs which means they'd have to auth separately if protected, a setup that's not that common so it that's probably where our blind spot is.

**Comment by sinbad** [comment]
Turns out `doStorageRequest` does look up credentials & supply them the same way. So maybe there's something in the headers which is stopping that, can you try setting `GIT_CURL_VERBOSE=1` as well to get even more detail?

**Comment by rjbell4** [comment]
Verbose log attached:  [fetch-verbose.txt](https://github.com/github/git-lfs/files/202910/fetch-verbose.txt)

I also managed to compile git-lfs in my environment, so I can make local changes, recompile and test them.  I just don't know my way through the source code.  I did add a printf to doStorageRequest, which you can see in the output, but that's it.  It also means that this more recent output is with the latest code.

**Comment by rjbell4** [comment]
I've been poking around, and instrumenting git-lfs to learn more about it.  My instrument log is attached: [fetch-verbose.txt](https://github.com/github/git-lfs/files/203205/fetch-verbose.txt)
Mostly, I just added trace lines and stack traces to learn more about what's going on.  You probably don't need it, but I did.

It seems like the code creates the "transfer queues" which operate in parallel before it ever attempts any HTTP requests of the server.  It seems the "main thread" does a series of Batch API request, which don't require credentials, apparently because my LFS URL isn't set for private Access.  And those HTTP requests succeeds.  The batch results are then queued up for the transfer workers.  Those workers attempt to pull objects from the LFS server, and call doStorageRequest and subsequently getCreds to fill in credentials.  Hence, I have many different threads (or "goroutines", as it seems they are called) each attempting to fill credentials concurrently, and it leads to the mess that I've reported.

I've tried setting lfs.<url>.access to ```private```, not because it's necessary, but with the idea that that would get the credentials up front, and then they wouldn't be needed later.  And indeed, that seemed to work -- the credentials were "cleanly" asked for up front, and the transfers succeeded with the cached authentication.

So I think I have a work-around.  However, it seems to me that it's just that -- a work-around.  Is it unique or problematic that the Batch API request wouldn't require credentials, but that they'd be used for the downloading of the objects?  If that's not an issue in and of itself, then it seems that I shouldn't be "forced" to set the access method to ```private``` to get usable credential prompting.

BTW, I was hoping to add lfs.<url>.access to ```.lfsconfig```, as that would be the smoothest way to support this (and it seems logical to define the URL as private the same place that I define the URL), but that doesn't appear to work.

**Comment by rjbell4** [comment]
Added PR #1122 to be able to specify the access type as private in .lfsconfig.  I thought we should treat that as a slightly separate issue, even though it spawned from this one.

**Comment by sinbad** [comment]
That makes sense - yeah it was clearly that the download threads were failing auth while the batch API call wasn't, it just wasn't clear why but the API endpoint not having any auth would explain it. This is actually the opposite of the way most implementations do it, it's more typical to put the auth on the API and then have the API generate a token which is used in the download (so that doesn't prompt for auth). It's not technically wrong, just unusual. The reason it makes more sense to do it on the Batch API call is:

1. Principle of validating early; responding to a batch request takes resources and can also expose *some* metadata about the repo (although just what objects are there)
1. The Batch API call usually includes context relating to the repository, which means you know the permissions on it; the actual store may have less context / de-duplicate and so granular permissions are higher up

Setting your local config to `private` makes the client look up credentials before calling the Batch API which causes you to get the prompt ahead of time so works around this. 

> Is it unique or problematic that the Batch API request wouldn't require credentials, but that they'd be used for the downloading of the objects?

Yeah, I haven't seen it done like that before; we don't force it to be done the other way although it does normally make more sense. 

A permanent fix for this kind of configuration would be to always issue the very first transfer request serially and therefore to allow a 401 from that to result in a single prompt, and to only continue with the other workers once a success response has been received from the first request. It does mean slightly slowing down the initial transfers but since the result code will come through before the whole body it should be doable. The question is whether we do that, or change the definition of the API to say that auth must be done on the Batch API if the storage calls require auth (which is less complex for the client). What do you think @technoweenie?

**Comment by sinbad** [comment]
Aside: I noticed some concurrent map errors in your output which I've fixed in https://github.com/github/git-lfs/commit/95272fef409fa604e766ba4cf5a7257630d9c1c1.

**Comment by technoweenie** [comment]
Ahhh, that makes sense. Thanks for the logs, @rjbell4.

I agree that LFS APIs should require auth upfront, while generating tokens in the object responses. Considering this is easy to workaround (setting lfs.access to `basic`), I'm okay with leaving the client behavior where it is for now, and hardening the docs as you suggested.

Still, it's not desirable behavior in the client. It'd be nice to fix, or at least _detect_. It's worth keeping around on the roadmap.
